Kotlin-sender burst: first link DATA packet post-establishment intermittently lost (kotlin→ref→ref)

## Summary

`reticulum-conformance/tests/wire/test_link_multihop.py::test_link_data_roundtrip_multiple_packets` intermittently fails in the **`kotlin→reference→reference`** parametrization: the receiver gets 4 out of 5 link DATA packets. The missing packet is always the **first** one in the burst.

Observed frequency: **~33% on a 3-run sample** (1 failure in 3 runs across two branch heads). Already costing a CI rerun on [PR #41](https://github.com/torlando-tech/reticulum-kt/pull/41).

## Topology

```
kotlin sender ──TCPClient──▶ reference transport (enable_transport=true) ◀──TCPClient── reference receiver
```

Both sender and receiver are TCPClients of the transport peer; all three in-process bridge subprocesses, all on loopback.

## Reproducer

```bash
cd reticulum-conformance
pytest tests/wire/test_link_multihop.py::test_link_data_roundtrip_multiple_packets -k 'kotlin->reference->reference' -v --count 10
```

(The `--count` needs pytest-repeat; without it, manually rerun until it flips red.)

## Failure signature

Exact assertion from CI ([run 24622089275](https://github.com/torlando-tech/reticulum-kt/actions/runs/24622089275/job/71994588312)):

```
AssertionError: receiver(reference) got 4 packets from sender(kotlin) (expected 5).
Missing: {b'f\xd2Y\xc6\xe3\xc2ivP\xf8\x85)\x85ml\x92'}. Extra: set().
```

**Missing payload is 16 bytes.** The test constructs payloads as:

```python
payloads = [secrets.token_bytes(16 + i * 8) for i in range(5)]
# → sizes 16, 24, 32, 40, 48
```

So the 16-byte payload is **`payloads[0]`** — the *first* packet sent after link establishment. This is a stable clue, not random: the inter-send gap is already 50ms and the poll timeout is 10s, so timing has plenty of slack.

## Narrowed hypothesis: ACTIVE-before-ready race

The sequence the test runs:

1. `sender.link_open(dest_hash, …)` — bridge command that blocks until `link.status == ACTIVE` or timeout.
2. Loop 5× over `sender.link_send(link_id, payload)` with `time.sleep(0.05)` between.

If the first payload is the one dropped, **the link reports ACTIVE before the outbound path is actually ready to transmit on it**. Candidate code paths worth inspecting:

- Post-LRPROOF handling in `Transport` / `Link`: does `Link.status` flip to `ACTIVE` *before* any bookkeeping needed for `Link.send() → Transport.outbound()` completes? (e.g., `activeLinks` registration, path/link-table entry for the responder direction, the `hadOutbound(isData=true)` sequencing state.)
- `Transport.registerLinkPath` → `pathTable[linkId]` write: is there a window where `Link.status == ACTIVE` but `pathTable` lookup in `processOutbound` still returns null, causing the first `Transport.outbound(packet)` to fall into the `!sent` broadcast fallback? The broadcast fallback is gated by the link's attached interface, which should still transmit — unless the `attachedInterfaceHash` getter on the link returns null during that same window.
- LRRTT packet emission right after ACTIVE: I see `[Transport] Outbound LINK packet: dest=<linkId>, context=LRRTT` logs immediately on establishment. If LRRTT and the first user DATA share a lock/queue and LRRTT's processing is synchronous-but-slow, the user DATA write might race — though this would usually delay, not drop.

## Related, distinct issue

`_xfail_kotlin_receiver_multihop` in the same test file already documents a **kotlin-as-*receiver*** multi-hop packet-loss bug (several `*→*→kotlin` cases are `xfail`'d). That's inbound. This new issue is the **kotlin-as-*sender*** mirror: outbound burst losing the first packet. The symptoms are similar (intermittent packet loss under burst) but the code paths are different (inbound handler dispatch vs. outbound post-ACTIVE emit).

## Run history (for flake rate baseline)

On branch `test/guard-link-data-header1-multihop` (which only adds a unit test file, no production changes):

| Run                                                                                                                          | kotlin→ref→ref burst | Other kot-sender cases |
| ---------------------------------------------------------------------------------------------------------------------------- | -------------------- | ---------------------- |
| [24621983166](https://github.com/torlando-tech/reticulum-kt/actions/runs/24621983166)                                        | PASS                 | PASS                   |
| [24622089275](https://github.com/torlando-tech/reticulum-kt/actions/runs/24622089275/job/71994588312)                        | FAIL (missing[0])    | PASS                   |
| [24622089275 rerun](https://github.com/torlando-tech/reticulum-kt/actions/runs/24622089275/job/71994781987)                  | PASS                 | PASS                   |

Single-packet variant `test_link_data_reaches_receiver_multihop[kotlin→ref→ref]` passes in all three runs — so the 1-hop-after-ACTIVE single send is clean; only the immediate burst is racy.

## What would close this

- Root-cause the race between `Link.status = ACTIVE` and the outbound path's readiness to transmit the *first* user DATA packet.
- Fix so that `link_open` does not return (or `Link.status` does not flip to ACTIVE) until `Transport.outbound(DATA)` would succeed on the first try.
- Once fixed, consider adding a reticulum-kt unit test that post-LRPROOF emits a DATA packet as the very next outbound and asserts it actually hits the wire (companion to [PR #41](https://github.com/torlando-tech/reticulum-kt/pull/41)'s HEADER_1-on-multihop guard).

🤖 Filed via [Claude Code](https://claude.com/claude-code)

--- a/rns-core/src/main/kotlin/network/reticulum/link/Link.kt
+++ b/rns-core/src/main/kotlin/network/reticulum/link/Link.kt
@@ -606,12 +606,21 @@ class Link private constructor(
                 return false
             }
 
-            // Link is now active
+            // Link is now active — but do NOT publish status=ACTIVE until every
+            // piece of bookkeeping that an outbound send() depends on is in place.
+            // `status` is @Volatile, so the final write at the end of this block
+            // acts as a release fence: any thread that subsequently observes
+            // status==ACTIVE is guaranteed to see the activeLinks membership,
+            // attachedInterfaceHash, and pathTable entry written earlier.
+            //
+            // This ordering fixes the race behind #42: bridge callers that
+            // `link_open(...)` and then immediately `link_send(first_payload)`
+            // could previously observe status=ACTIVE during the ~16-line window
+            // before registerLinkPath ran, causing Transport.outbound() to miss
+            // the pathTable entry and drop the first DATA packet.
             rtt = System.currentTimeMillis() - requestTime
             remoteIdentity = destination.identity
             mdu = LinkConstants.calculateMdu(mtu)
-            status = LinkConstants.ACTIVE
-            activatedAt = System.currentTimeMillis()
 
             // Calculate establishment rate (bytes per ms)
             val linkRtt = rtt
@@ -628,6 +637,10 @@ class Link private constructor(
             packet.receivingInterfaceHash?.let { interfaceHash ->
                 Transport.registerLinkPath(linkId, interfaceHash, packet.hops)
             }
+
+            // Publish ACTIVE only after all the above is visible.
+            activatedAt = System.currentTimeMillis()
+            status = LinkConstants.ACTIVE
 
             log("Link ${linkId.toHexString()} established, RTT: ${rtt}ms")
 

--- a/rns-core/src/main/kotlin/network/reticulum/link/Link.kt
+++ b/rns-core/src/main/kotlin/network/reticulum/link/Link.kt
@@ -1886,8 +1886,12 @@ class Link private constructor(
 
             // Use max of measured and remote RTT
             rtt = maxOf(measuredRtt, remoteRtt)
-            status = LinkConstants.ACTIVE
+            // Set activatedAt BEFORE the volatile status write so any thread that
+            // observes status==ACTIVE via the volatile read is also guaranteed to
+            // see a non-zero activatedAt (getAge()/noInboundFor() correctness).
+            // This mirrors the ordering applied to validateProof() for issue #42.
             activatedAt = System.currentTimeMillis()
+            status = LinkConstants.ACTIVE
 
             // Calculate establishment rate (bytes per ms)
             val linkRtt = rtt

